### PR TITLE
Basic bandits implementation

### PR DESCRIPTION
**Pull request by koihia**

### CODE REVIEW

File: src/main/java/org/elasql/perf/tpart/SpCallPreprocessor.java [hunk 1]
```diff
@@ -77,13 +82,18 @@ public SpCallPreprocessor(TPartStoredProcedureFactory factory,
 			dependencyRecorder = new TransactionDependencyRecorder();
 			dependencyRecorder.startRecording();
 		}
+
+		if (Elasql.SERVICE_TYPE == Elasql.ServiceType.HERMES_BANDIT) {
+			BanditBasedRouter banditBasedRouter = (BanditBasedRouter) inserter;
```
Comment (SLMT) [inline src/main/java/org/elasql/perf/tpart/SpCallPreprocessor.java:87]: It seems like `banditBasedRouter` is not used in any part of code. What does this line do?

File: src/main/java/org/elasql/remote/groupcomm/server/ConnectionMgr.java [hunk 2]
```diff
@@ -103,6 +104,11 @@ public void sendMetricReport(MetricReport report) {
 		commServer.sendP2pMessage(ProcessType.SERVER, SEQUENCER_ID, report);
 	}
 
+	public void sendTransactionMetricReport(TransactionMetricReport report) {
+		// XXX: is vanilacomm thread safe?
```
Comment (SLMT) [inline src/main/java/org/elasql/remote/groupcomm/server/ConnectionMgr.java:108]: Yes, it's thread-safe.

**Review by SLMT — APPROVED** [review]
Great work! I am glad to see that you strictly follow our architecture when you implemented the bandit models. I also see that you carefully modularized your code and wrote comments to make your code easier to understand. This is definitely a well written implementation.

Here are some suggestions:
- I see that you add `TransactionMetricReport` to send reward back to the sequencer. However, we already have `CommitNotification` that does similar job. You may add transaction latency to `CommitNotification` and calculate reward in `TPartPerformanceManager` when it receives the notification. This can reduce many duplicate code in your implementation.
- I am also suggesting that you may consider the design of putting the bandit model in the sequencer since I notice that you send lots of data for routing (context) and updating the model (arms and rewards). Please contact @xuan19970825 for more information.

There are also a few comments in the code. Don't forget to check them.

Well done.

File: src/main/java/org/elasql/perf/tpart/SpCallPreprocessor.java (hunk 1 above)
Comment (koihia) [reply src/main/java/org/elasql/perf/tpart/SpCallPreprocessor.java:87]: It does nothing. It should have been cleaned up.

**Pull request merged**